src/ambianic/pipeline/ai/tf_detect.py

```diff
@@ -0,0 +1,180 @@
+"""Tensorflow image detection wrapper."""
+import logging
+import time
+import re
+import numpy as np
+# from importlib import import_module
+from PIL import ImageOps
+from .inference import TFInferenceEngine
+from ambianic.pipeline import PipeElement
+
+
+log = logging.getLogger(__name__)
+
+
+class TFDetectionModel(PipeElement):
+    """Applies Tensorflow image detection."""
+
+    def __init__(self,
+                 model=None,
+                 labels=None,
+                 label_filter=None,
+                 confidence_threshold=0.6,
+                 top_k=3,
+                 **kwargs
+                 ):
+        """Initialize detector with config parameters.
+        :Parameters:
+        ----------
+        model: ai_models/mobilenet_ssd_v2_face.tflite
+        labels: ai_models/coco_labels.txt
+        confidence_threshold: 0.6
+        top_k: 3
+        """
+
+        # log.warning('TFImageDetection __init__ invoked')
+        super().__init__(**kwargs)
+
+        self._tfengine = TFInferenceEngine(
+            model=model,
+            labels=labels,
+            confidence_threshold=confidence_threshold,
+            top_k=top_k)
+        self._labels = self.load_labels(self._tfengine.labels_path)
+        self._label_filter = label_filter
+        self.last_time = time.monotonic()
+
+    def load_labels(self, label_path=None):
+        """Load label mapping from integer code to text.
+        :Parameters:
+        ----------
+        label_path : string
+            Path to label mapping file.
+        :Returns:
+        -------
+        dict
+            {label_code, label_text}
+        """
+        assert label_path
+        p = re.compile(r'\s*(\d+)(.+)')
+        with open(label_path, 'r', encoding='utf-8') as f:
+            lines = (p.match(line).groups() for line in f.readlines())
+            return {int(num): text.strip() for num, text in lines}
+
+    @staticmethod
+    def thumbnail(image=None, desired_size=None):
+        """Resizes original image as close as possible to desired size.
+        Preserves aspect ratio of original image.
+        Does not modify the original image.
+        :Parameters:
+        ----------
+        image : PIL.Image
+            Input Image for AI model detection.
+        desired_size : (width, height)
+            Size expected by the AI model.
+        :Returns:
+        -------
+        PIL.Image
+            Resized image fitting for the AI model input tensor.
+        """
+        assert image
+        assert desired_size
+        log.debug('input image size = %r', image.size)
+        thumb = image.copy()
+        w, h = desired_size
+        try:
+            # convert from numpy to native Python int type
+            # that PIL expects
+            if isinstance(w, np.generic):
+                w = w.item()
+                w = int(w)
+                h = h.item()
+                h = int(h)
+            thumb.thumbnail((w, h))
+        except Exception as e:
```

Comment:
@bhavikapanara could you please add unit test coverage for the lines the codecov flagged?

Reply:
@ivelin I didn't get any idea what exception case should I generate for this function. Please guide me.

Reply:
Receiving meaningless parameters should cause an exception.

Reply:
@ivelin In this function there is only two parameters that are ` w, h = desired_size` and these value w,h are coming from [TensorFlow input model parameter ](https://github.com/ambianic/ambianic-edge/blob/80e1b367c49e8ba32289b9f805364518962e2886/src/ambianic/pipeline/ai/image_boundingBox_detection.py#L113)so that can't be incorrect. Please take a look at it.

```
height = tfe.input_details[0]['shape'][1]
width = tfe.input_details[0]['shape'][2]

desired_size = (width, height)
thumbnail = TFDetectionModel.thumbnail(image=image, desired_size=desired_size)
```

Reply:
Right. The purpose of the test is to force an exception just in case this ever happens at runtime. Imagine there is a bug in the tfmodel or it outputs crazy tensor values.

Reply:
Done